### PR TITLE
Anchor pallet

### PR DESCRIPTION
**Pull request by drewstone**
This PR adds a lot of architectural functionality modelled after: https://github.com/webb-tools/darkwebb-solidity/tree/main/contracts/anchors/bridged.

Primarily it integrates the Anchor pallet. This pallets:
- Consumes a verifier and a mixer module
- Exposes a deposit & withdraw API modelled after our solidity implementation
- Exposes a graph-like API for adding edges to trees, denoted by a map from tree_id -> chain_id -> edge metadata

Tests will be added in future PRs. For now, this gets the tests compiling for the other modules and compiling for the repo at large.

### CODE REVIEW

**Review by dev0x1 — CHANGES_REQUESTED** [review]

File: pallets/mixer/src/lib.rs [hunk 1]
```diff
@@ -200,15 +222,28 @@ impl<T: Config<I>, I: 'static> MixerInterface<T, I> for Pallet<T, I> {
...
 		relayer: T::AccountId,
 		fee: BalanceOf<T, I>,
+		refund: BalanceOf<T, I>,
 	) -> Result<(), DispatchError> {
-		let root = T::Tree::get_root(id)?;
+		// Check if local root is known
+		ensure!(T::Tree::is_known_root(id, roots[0])?, Error::<T, I>::InvalidWithdrawRoot);
+		// Check nullifier and add or return `InvalidNullifier`
+		ensure!(
+			<Self as MixerInspector<_,_>>::is_nullifier_used(id, nullifier_hash),
```
[the remark this answers is not in the document]
Comment (drewstone) [reply pallets/mixer/src/lib.rs:236]: Right, good catch!
```suggestion
			!<Self as MixerInspector<_,_>>::is_nullifier_used(id, nullifier_hash),
```

File: pallets/anchor/src/types.rs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+		src_chain_id: T::ChainId,
+		root: T::Element,
+		height: T::BlockNumber
+	) -> Result<(), dispatch::DispatchError>;
+}
+
+/// Anchor trait for inspecting tree state
+pub trait AnchorInspector<T: Config<I>, I: 'static = ()> {
+	/// Gets the merkle root for a tree or returns `TreeDoesntExist`
+	fn get_neighbor_roots(id: T::TreeId) -> Result<Vec<T::Element>, dispatch::DispatchError>;	/// Checks if a merkle root is in a tree's cached history or returns
```
[the remark this answers is not in the document]
Comment (drewstone) [reply pallets/anchor/src/types.rs:46]: Probably line break needed, will check again.

File: pallets/mixer/src/lib.rs [hunk 3]
```diff
@@ -181,6 +199,10 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 }
 
 impl<T: Config<I>, I: 'static> MixerInterface<T, I> for Pallet<T, I> {
+	fn create(creator: T::AccountId, depth: u8) -> Result<T::TreeId, DispatchError> {
+		T::Tree::create(T::AccountId::default(), 32u8)
```
[the remark this answers is not in the document]
Comment (drewstone) [reply pallets/mixer/src/lib.rs:203]: I think I rather meant to use the passed in depth here.
```suggestion
		T::Tree::create(T::AccountId::default(), depth)
```

File: pallets/verifier/Cargo.toml [hunk 4]
```diff
@@ -31,6 +31,7 @@ sp-core = { default-features = false, version = "3.0.0", git = "https://github.c
 sp-io = { default-features = false, version = "3.0.0", git = "https://github.com/paritytech/substrate.git", branch = "polkadot-v0.9.7" }
 sp-runtime = { default-features = false, version = "3.0.0", git = "https://github.com/paritytech/substrate.git", branch = "polkadot-v0.9.7" }
 pallet-balances = { default-features = false, version = "3.0.0", git = "https://github.com/paritytech/substrate.git", branch = "polkadot-v0.9.7" }
+ark-crypto-primitives = { version = "^0.3.0", features = ["r1cs"], default-features = false }
```
[the remark this answers is not in the document]
Comment (drewstone) [reply pallets/verifier/Cargo.toml:34]: Perhaps but it is likely we will stick with these SNARKs for now.

**Review by dev0x1 — APPROVED** [review]

**Review by shekohex — APPROVED** [review]

**Pull request merged**